src/lib/y2partitioner/actions/controllers/disk_device.rb

```diff
@@ -0,0 +1,199 @@
+        #   also avoided.
+        #
+        # @param device [Y2Storage::Partitionable] device where to clone
+        # @return [Boolean]
+        def suitable_for_cloning?(device)
+          return false if self_cloning?(device)
+
+          suitable =
+            enough_size_for_cloning?(device) &&
+            support_partition_table_type?(device) &&
+            same_topology?(device)
+
+          Yast::Mode.installation ? suitable : suitable && !mount_points?(device)
+        end
+
+        # Whether it is trying a self cloning
+        #
+        # @note A device cannot be self cloned.
+        #
+        # @param device [Y2Storage::Partitionable] device where to clone
+        # @return [Boolean]
+        def self_cloning?(device)
+          device.sid == self.device.sid
+        end
+
+        # Whether the given device has enough size for the cloning
+        #
+        # @param device [Y2Storage::Partitionable] device where to clone
+        # @return [Boolean]
+        def enough_size_for_cloning?(device)
+          device.size >= self.device.size
+        end
+
+        # Whether the given device supports the partition table type of the current device
+        #
+        # @param device [Y2Storage::Partitionable] device where to clone
+        # @return [Boolean]
+        def support_partition_table_type?(device)
+          device.possible_partition_table_types.include?(self.device.partition_table.type)
+        end
+
+        # Whether the given device has the same topology than the current device
+        #
+        # @param device [Y2Storage::Partitionable] device where to clone
+        # @return [Boolean]
+        def same_topology?(device)
+          device.topology == self.device.topology
+        end
+
+        # Whether the given device holds mount points
+        #
+        # @note If something mounted depends on the given device, that device cannot be
+        #   used for cloning. Firstly, mount points should be removed.
+        #
+        # @param device [Y2Storage::Partitionable] device where to clone
+        # @return [Boolean]
+        def mount_points?(device)
+          device.descendants.any? { |d| d.is?(:mount_point) }
+        end
+
+        # Wipes the given device
+        #
+        # @note All its descendats are removed.
+        #
+        # @param device [Y2Storage::BlkDevice]
+        def wipe_device(device)
+          device.remove_descendants
+        end
+
+        # Clones the partition table of the current device into the given device
+        #
+        # @note All partitions of the current device are cloned into the given
+        #   device, see {#clone_partition}.
+        #
+        # @param device [Y2Storage::Partitionable] device where to clone
+        def clone_partition_table(device)
+          device.create_partition_table(self.device.partition_table.type)
+          sorted_partitions = self.device.partitions.sort_by(&:name)
```

Comment:
why self is needed?

Reply:
`Controllers::DiskDevice#device` is the device associated to the controller. In the method we have also a param `device`. We need self here to take the partitions from the device associated to the controller and not from the param device.

Reply:
ok, I suggest to try nicer name here to avoid errors when modifying this method. What about param name `partitionable`?

Reply:
In general, we are avoiding `partitionable` concept, so I would not use that name. I can change it to `target_device`.